File: sdks/typescript/src/resources/drafts.ts
// Drafts Resource for OpenAgentMail

import type { HttpClient } from '../utils/http.js';
import { paginate, PageIterator } from '../utils/pagination.js';
import type {
  Draft,
  Message,
  CreateDraftParams,
  UpdateDraftParams,
  PaginatedResponse,
  PaginationParams,
} from '../types.js';

/**
 * Client for managing drafts
 */
export class DraftsClient {
  constructor(private readonly http: HttpClient) {}

  /**
   * Create a new draft
   */
  async create(inboxId: string, params: CreateDraftParams = {}): Promise<Draft> {
    return this.http.post<Draft>(`/inboxes/${inboxId}/drafts`, params);
  }

  /**
   * Get a draft by ID
   */
  async get(inboxId: string, draftId: string): Promise<Draft> {
    return this.http.get<Draft>(`/inboxes/${inboxId}/drafts/${draftId}`);
  }

  /**
   * Update a draft
   */
  async update(inboxId: string, draftId: string, params: UpdateDraftParams): Promise<Draft> {
    return this.http.patch<Draft>(`/inboxes/${inboxId}/drafts/${draftId}`, params);
  }

  /**
   * Delete a draft
   */
  async delete(inboxId: string, draftId: string): Promise<void> {
    return this.http.delete(`/inboxes/${inboxId}/drafts/${draftId}`);
  }

  /**
   * Send a draft immediately
   */
  async send(inboxId: string, draftId: string): Promise<Message> {
    return this.http.post<Message>(`/inboxes/${inboxId}/drafts/${draftId}/send`);
  }

  /**
   * List drafts in an inbox with auto-pagination
   */
  list(inboxId: string, params: PaginationParams = {}): PageIterator<Draft> {
    return paginate(
      (paginationParams) => this.listPage(inboxId, { ...params, ...paginationParams }),
      params
    );
  }

  /**
   * List a single page of drafts
   */
  async listPage(inboxId: string, params: PaginationParams = {}): Promise<PaginatedResponse<Draft>> {
    const { limit, pageToken } = params;
    return this.http.get<PaginatedResponse<Draft>>(`/inboxes/${inboxId}/drafts`, {
      limit,
      pageToken,
    });
  }
}

File: sdks/rust/src/resources/drafts.rs
use std::sync::Arc;
use crate::client::ClientInner;
use crate::error::Result;
use crate::types::{CreateDraftRequest, Draft, Message, PaginatedResponse, PaginationParams, UpdateDraftRequest};

#[derive(Debug, Clone)]
pub struct Drafts { inner: Arc<ClientInner> }

impl Drafts {
    pub(crate) fn new(inner: Arc<ClientInner>) -> Self { Self { inner } }

    pub async fn create(&self, inbox_id: &str, request: CreateDraftRequest) -> Result<Draft> {
        let response = self.inner.post(&format!("/inboxes/{}/drafts", inbox_id)).json(&request).send().await?;
        self.inner.handle_response(response).await
    }

    pub async fn list(&self, inbox_id: &str, params: PaginationParams) -> Result<PaginatedResponse<Draft>> {
        let mut request = self.inner.get(&format!("/inboxes/{}/drafts", inbox_id));
        for (key, value) in params.to_query_params() { request = request.query(&[(key, value)]); }
        let response = request.send().await?;
        self.inner.handle_response(response).await
    }

    pub async fn get(&self, inbox_id: &str, draft_id: &str) -> Result<Draft> {
        let response = self.inner.get(&format!("/inboxes/{}/drafts/{}", inbox_id, draft_id)).send().await?;
        self.inner.handle_response(response).await
    }

    pub async fn update(&self, inbox_id: &str, draft_id: &str, request: UpdateDraftRequest) -> Result<Draft> {
        let response = self.inner.patch(&format!("/inboxes/{}/drafts/{}", inbox_id, draft_id)).json(&request).send().await?;
        self.inner.handle_response(response).await
    }

    pub async fn send(&self, inbox_id: &str, draft_id: &str) -> Result<Message> {
        let response = self.inner.post(&format!("/inboxes/{}/drafts/{}/send", inbox_id, draft_id)).send().await?;
        self.inner.handle_response(response).await
    }

    pub async fn delete(&self, inbox_id: &str, draft_id: &str) -> Result<()> {
        let response = self.inner.delete(&format!("/inboxes/{}/drafts/{}", inbox_id, draft_id)).send().await?;
        self.inner.handle_empty_response(response).await
    }
}
